core-backend/src/funcs.rs

```diff
@@ -1297,21 +1298,36 @@ where
         })
     }
 
-    pub fn out_of_gas(_gas: Gas) -> impl Syscall<Ext> {
-        RawSyscall::new(|ctx: &mut CallerWrap<Ext>| {
-            let ext = ctx.ext_mut();
-            let current_counter = ext.current_counter_type();
-            log::trace!(target: "syscalls", "[out_of_gas] Current counter in global represents {current_counter:?}");
+    fn out_of_gas(ctx: &mut CallerWrap<Ext>) -> UndefinedTerminationReason {
+        let ext = ctx.ext_mut();
+        let current_counter = ext.current_counter_type();
+        log::trace!(target: "syscalls", "system_break(OutOfGas): Current counter in global represents {current_counter:?}");
 
-            if current_counter == CounterType::GasAllowance {
-                // We manually decrease it to 0 because global won't be affected
-                // since it didn't pass comparison to argument of `gas_charge()`
-                ext.decrease_current_counter_to(0);
-            }
+        if current_counter == CounterType::GasAllowance {
+            // We manually decrease it to 0 because global won't be affected
+            // since it didn't pass comparison to argument of `gas_charge()`
+            ext.decrease_current_counter_to(0);
+        }
+
+        ActorTerminationReason::from(current_counter).into()
+    }
 
-            let termination_reason: ActorTerminationReason = current_counter.into();
+    fn stack_limit_exceeded() -> UndefinedTerminationReason {
+        log::trace!(target: "syscalls", "system_break(StackLimitExceeded)");
+        TrapExplanation::StackLimitExceeded.into()
+    }
+
+    pub fn system_break(_gas: Gas, code: u64) -> impl Syscall<Ext> {
+        RawSyscall::new(move |ctx: &mut CallerWrap<Ext>| {
+            let termination_reason = match SystemBreakCode::try_from(code) {
+                Ok(system_break_code) => match system_break_code {
+                    SystemBreakCode::OutOfGas => Self::out_of_gas(ctx),
+                    SystemBreakCode::StackLimitExceeded => Self::stack_limit_exceeded(),
+                },
+                _ => TrapExplanation::Unknown.into(),
```

Comment:
aren't we supposed to panic here?
its only case if we failed up instrumentation process, isn't it? @grishasobol

Reply:
In fact, this is an unreachable branch, because we never use codes other than `0` and `1` at instrumentation level. Do you mean do `_ => unreachable!("msg")` instead of `Trap(Unknown)`?

Reply:
I think `Trap(Unknown)` compiles to smaller binary compared to panic.

Reply:
> In fact, this is an unreachable branch, because we never use codes other than `0` and `1` at instrumentation level. Do you mean do `_ => unreachable!("msg")` instead of `Trap(Unknown)`?

Yes

Reply:
> I think `Trap(Unknown)` compiles to smaller binary compared to panic. The `free` syscall uses the same approach instead of panicking.

But free can be called by program in its initial logic, while this call is only created already inside our runtime.

Reply:
done in 3632f47607bee0bcb622b7bf088b36a26d2d33bb